bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+import pandas as pd
+from event_model import DocumentRouter, RunRouter
+from tiled.client import from_profile, from_uri
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.node = None
+        self._descriptor_nodes = {}  # references to descriptor containers by uid's
+        self._SR_nodes = {}
+        self._SR_cache = {}
+
+    def start(self, doc):
+        self.node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        metadata = dict(self.node.metadata) | {"stop": doc}
+        self.node.update_metadata(metadata=metadata)
+
+    def descriptor(self, doc):
+        descriptor_node = self.node.create_container(key=doc["name"], metadata=doc)
+        self._descriptor_nodes[doc["uid"]] = descriptor_node
+        descriptor_node.create_container(key="external")
+        descriptor_node.create_container(key="internal")
+
+    def event(self, doc):
+        descriptor_node = self._descriptor_nodes[doc["descriptor"]]
+        parent_node = descriptor_node["internal"]
+        for table_key in ["data", "timestamps"]:
```

Comment:
I'll add a note here as a reminder to ourselves: we discussed making a single table here (not sure what to call it...`"table"`?) that contains both the data and timestmaps columns, potentially prefixed like `data_*` and `ts_*` as a way of namespacing them.

Reply:
This is something I was going to add last Friday just before I was caught up with flyscanning at the lab. Added now.

Reply:
I have named the entire table `/internal/data`, and the column names are taken directly from the corresponding data keys: `{key1}, {key2}, ..., ts_{key1}, ts_{key2}, ...`. This seems the most compact to me, but happy to change it to another schema.

Reply:
I think that's probably best. We technically have a problem if someone names a field `ts_`, but I think this is a time to be pragmatic and say, "Don't do that!" We can disallow it in the schema if we want to.